### PR TITLE
Close RocksDB objects that are AutoCloseable

### PR DESCRIPTION
**Pull request by jiacheliu3**
### What changes are proposed in this pull request?

Fixes https://github.com/facebook/rocksdb/issues/9378
The `orgs.rocksdb.RocksObject` extends `AutoCloseable` so they should be closed properly to avoid leaks. 

### Why are the changes needed?

According to the [RocksDB documentation](https://github.com/facebook/rocksdb/wiki/RocksJava-Basics#memory-management) and a sample [PR discussion](https://github.com/facebook/rocksdb/pull/7608#discussion_r515159270), the objects need to be closed or wrapped in a try-with-resource block.

### Does this PR introduce any user facing changes?

NA

### CODE REVIEW

**Comment by jiacheliu3** [comment]
@apc999 @yuzhu This is a WIP fix which closes the RocksObject instances in a better way.

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java [hunk 1]
```diff
@@ -161,18 +157,28 @@ public long size() {
...
-    // Explicitly hold a reference to the ReadOptions object from the discussion in
+    // Explicitly hold a reference to the objects from the discussion in
     // https://groups.google.com/g/rocksdb/c/PwapmWwyBbc/m/ecl7oW3AAgAJ
-    final ReadOptions readOptions = new ReadOptions().setIterateUpperBound(new Slice(endKey));
-    try (RocksIterator iter = db().newIterator(mBlockLocationsColumn.get(), readOptions)) {
+    try (
+         // TODO(jiacheng): why final?
+         final Slice slice = new Slice(endKey);
+         final ReadOptions readOptions = new ReadOptions().setIterateUpperBound(slice);
+         final RocksIterator iter = db().newIterator(mBlockLocationsColumn.get(), readOptions)) {
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java:181]: According to the Oracle doc, the close order is opposite to the declaration order
https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/RocksPageStore.java [hunk 2]
```diff
@@ -81,17 +83,18 @@ public static RocksPageStore open(RocksPageStoreOptions options) throws IOExcept
       }
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/cache/store/RocksPageStore.java:83]: close option too before `}` just to be complete--- but I understand it matter little here as we are throwing IOE

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java [hunk 3]
```diff
@@ -74,18 +75,12 @@ public RocksBlockStore(String baseDir) {
     mDisableWAL = new WriteOptions().setDisableWAL(true);
     mIteratorOption = new ReadOptions().setReadaheadSize(
         ServerConfiguration.getBytes(PropertyKey.MASTER_METASTORE_ITERATOR_READAHEAD_SIZE));
-    ColumnFamilyOptions cfOpts = new ColumnFamilyOptions()
-        .setMemTableConfig(new HashLinkedListMemTableConfig())
-        .setCompressionType(CompressionType.NO_COMPRESSION);
+    mCfOpts = new ColumnFamilyOptions()
+            .setMemTableConfig(new HashLinkedListMemTableConfig())
+            .setCompressionType(CompressionType.NO_COMPRESSION);
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java:80]: nit: why indentation changed here?

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java [hunk 4]
```diff
@@ -96,7 +91,7 @@ public RocksBlockStore(String baseDir) {
         LOG.warn("Failed to create nonexistent db path at: {}. Error:{}", dbPath, e);
       }
     }
-    mRocksStore = new RocksStore(dbPath, backupPath, columns, dbOpts,
+    mRocksStore = new RocksStore("BlockStore", dbPath, backupPath, columns,
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java:94]: nit: better to define and reference a string constant

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java (hunk 1 above)
Comment (apc999) [reply core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java:180]: javadoc your clarification here?

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java [hunk 5]
```diff
@@ -58,8 +57,10 @@
   // This is a field instead of a constant because it depends on the call to RocksDB.loadLibrary().
   private final WriteOptions mDisableWAL;
   private final ReadOptions mIteratorOption;
+  private final ColumnFamilyOptions mCfOpts;
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java:60]: nit: let's keep naming more expressive
```suggestion
  private final ColumnFamilyOptions mColumnFamilyOptions;
```

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java [hunk 6]
```diff
@@ -50,32 +50,38 @@
...
-  public RocksStore(String dbPath, String checkpointPath,
-      Collection<ColumnFamilyDescriptor> columnFamilyDescriptors, DBOptions dbOpts,
+  public RocksStore(String name, String dbPath, String checkpointPath,
+      Collection<ColumnFamilyDescriptor> columnFamilyDescriptors,
       List<AtomicReference<ColumnFamilyHandle>> columnHandles) {
     Preconditions.checkState(columnFamilyDescriptors.size() == columnHandles.size());
+    mName = name;
     mDbPath = dbPath;
     mDbCheckpointPath = checkpointPath;
     mColumnFamilyDescriptors = columnFamilyDescriptors;
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java:78]: shall we close all members in this list?

**Review by apc999 — COMMENTED** [review]
thanks for the improvement. some comments left

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java [hunk 7]
```diff
@@ -217,6 +226,8 @@ public synchronized void restoreFromCheckpoint(CheckpointInputStream input) thro
   @Override
   public synchronized void close() {
     stopDb();
+    mDbOpts.close();
+    mDbOpts = null;
```
Comment (yuzhu) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java:230]: My personal preference is to never set a member variable to null unless it is passed in from outside of the object.  It was usually done to prevent obsolete references, but how could this reference be obsolete when this object crated it in the first place..

**Review by yuzhu — COMMENTED** [review]
Left a comment, thanks @jiacheliu3

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java (hunk 7 above)
Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java:230]: Oracle doc actually mentioned this https://docs.oracle.com/cd/E19159-01/819-3681/abebi/index.html

Setting the var to null helps the GC to tell that this object can be recycled. On 2nd thought in this particular case the `RocksObject` instances typically have the same life cycle as the holder instances themselves. So I don't see need in setting these to null explicitly. 

I will remove these setting to null lines. Also I don't have to remove the `final` and start worrying about thread safety again.

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java (hunk 6 above)
Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java:78]: I checked all objects from `org.rocksdb` and those objects I left alone do not inherit `RocksObject`, which means they are not `AutoCloseable`.

Discussed with @yuzhu today too, it's tricky to tell if an object is `AutoCloseable` and it's even harder to write UTs or detectors to ensure...

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/RocksPageStore.java (hunk 2 above)
Comment (jiacheliu3) [reply core/client/fs/src/main/java/alluxio/client/file/cache/store/RocksPageStore.java:83]: nice catch

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java (hunk 1 above)
Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java:180]: I believe you mean comment

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java [hunk 8]
```diff
@@ -208,7 +215,9 @@ public void removeLocation(long blockId, long workerId) {
 
   @Override
   public Iterator<Block> iterator() {
-    return RocksUtils.createIterator(db().newIterator(mBlockMetaColumn.get(), mIteratorOption),
+    // TODO(jiacheng): close the iterator when we iterate the BlockStore for backup
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java:218]: this will be done in a separate PR as it takes a bit refactoring. Now the leak only happens in backing up.

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java [hunk 9]
```diff
@@ -118,6 +118,7 @@ public boolean hasNext() {
       }
 
       @Override
+      // TODO(jiacheng): close this iterator properly on finish
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java:121]: tbd in the same separate PR

**Comment by jiacheliu3** [comment]
No test failures, what are the odds?

**Review by yuzhu — APPROVED** [review]

**Comment by yuzhu** [comment]
alluxio-bot, merge this please

**Comment by jiacheliu3** [comment]
alluxio-bot, cherry-pick to branch-2.7 please

**Comment by jiacheliu3** [comment]
alluxio-bot, cherry-pick this to branch-2.7 please
